File: 프로그래머스/2/389480. 완전범죄/완전범죄.js

function solution(info, n, m) {
  const memo = new Map();
  let answer = -1

  const dfs = (depth, a, b) => {
    if(answer > a) return  
    if (a <= 0 || b <= 0) return;
      
    const key = `${depth}-${a}-${b}`;
    if (memo.has(key)) return; 
    memo.set(key, true);
      
    if (depth === info.length) {
       answer = Math.max(answer, a)
       return
    } else {
      dfs(depth + 1, a - info[depth][0], b);
      dfs(depth + 1, a, b - info[depth][1]);  
      }
  };

  dfs(0, n, m);
  return answer < 0 ? - 1: n - answer
}

File: 프로그래머스/2/389480. 완전범죄/완전범죄.java
import java.util.*;


class Solution {
    int[][] info;
    int n;
    int m;
    int answer;
    Map<String, Boolean> visited = new HashMap<>();
        
    public int solution(int[][] info, int n, int m) {
        this.info = info;
        this.n = n;
        this.m = m;
        
        answer = -1;
        
        dfs(0, n, m);
        
        return answer < 0 ? -1 : n - answer;
    }
    
    void dfs(int idx, int aTrace, int bTrace) {
         if (aTrace <= 0 || bTrace <= 0) {
            return;
        }
        
        if (idx == info.length) { 
            answer = Math.max(answer, aTrace);
            return;
        }
        
        String key = idx + "," + aTrace + "," + bTrace;
        
         if (visited.containsKey(key)) {
            return;
        }
        
        visited.put(key, true);
        
        dfs(idx + 1, aTrace, bTrace - info[idx][1]);
        dfs(idx + 1, aTrace - info[idx][0], bTrace); 
    }
}